File: src/cmcandy/solutions/_0033.java
package cmcandy.solutions;

import javax.xml.stream.events.StartDocument;
import java.awt.*;

/**
 * Created by CHEN on 2020/5/8.
 */
public class _0033 {
    static class Solution {
        public int search(int[] nums, int target) {
            if (nums.length==0)return  -1;
            if (nums.length==1) return nums[0]==target?0:-1;
            int start = 0;
            int end = nums.length-1;
            while(start<end-1){
                int mid = start+(end-start)/2;
                if(nums[mid]<nums[start]) end = mid;
                else if(nums[mid]>=nums[start]) start = mid;
            }
            //获得了节点
            if (nums[start]<nums[end]){
                if (nums[0]>target||nums[nums.length-1]<target)return -1;
                return dicsearch(nums,0, nums.length-1,target);
            }

            else if(nums[start]<target||nums[end]>target) return -1;
            else if(nums[0]<=target)return dicsearch(nums,0,start,target);
            else if (nums[nums.length-1]>=target)return dicsearch(nums,end,nums.length-1,target);
            else return -1;
        }

        public int dicsearch(int[] nums,int first,int last,int target){
            if(first == last-1){
                if(nums[last]>target&&nums[first]<target)return -1;
                else if (nums[last]==target) return last;
                else if (nums[first]==target) return first;

            }
            int mid  = first+(last-first)/2;
            if(nums[mid] == target) return mid;
            else if (nums[mid]>target) return dicsearch(nums,first,mid,target);
            else if (nums[mid]<target)return dicsearch(nums,mid,last,target);
            return  -1;
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        System.out.println(solution.search(new int[]{4,5,6,7,0,1,2},0) );
        System.out.println(solution.search(new int[]{4,5,6,7,0,1,2},3) );
        System.out.println(solution.search(new int[]{1,3},1) );
        System.out.println(solution.search(new int[]{1,3,5},1) );

    }

}

File: src/cmcandy/Python_language_Answers/_0033.py
from typing import List



class Solution:
    def search(self, nums: List[int], target: int) -> int:
        left = 0
        right = len(nums) - 1
        if right == left:
            return 0 if nums[0] == target else -1
        if nums[left] < nums[right]:
            # 这里就是没有旋转，直接用二分就行
            return self.Binary(nums, left, right, target)

        jud = 1 if nums[0] <= target else 0

        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] < target:
                #这里=也可以，应该是测试样例没触碰到这里的边界 if jud == 1 and nums[mid] < nums[0]:
                if jud == 1 and nums[mid] <= nums[0]:
                    right = mid - 1
                else:
                    left = mid + 1
            elif nums[mid] > target:
                if jud == 0 and nums[mid] >= nums[0]:
                    left = mid + 1
                else:
                    right = mid - 1

            if nums[left] <= nums[right]:
                return self.Binary(nums, left, right, target)
        return -1

    def Binary(self, nums, left, right, target):
        while left <= right:
            mid = (left + right) // 2
            if nums[mid] == target:
                return mid
            elif nums[mid] > target:
                right = mid - 1
            elif nums[mid] < target:
                left = mid + 1
        return -1

if __name__ == '__main__':
    pass
    s = Solution()
    # print(s.search([4, 5, 6, 7, 0, 1, 2], 0))
    print(s.search([3, 1], 1))
    print(s.search([4, 5, 6, 7, 8, 1, 2, 3], 8))
    print(s.search([5, 1, 3], 3))
    print(s.search([5, 1, 2, 3, 4], 1))
    print(s.search([3, 5, 1], 3))
